### PR TITLE
[Dask] Add examples of loading multiple files into cube(s)

### PR DESCRIPTION
**Pull request by kaedonkers**
Added a section to notebook about "Using Iris Load with Dask".
This section contains examples of:
- Loading multiple files into multiple cubes
- Loading multiple files into a single cube

TODO:
- Load files into notebook from iris-sample-data library
- Loading multiple files with constraints

### CODE REVIEW

File: course_content/notebooks/dask_intro.ipynb [hunk 1]
```diff
@@ -393,6 +393,281 @@
...
+    "type(bag_cubelist)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It is worth noting that part of the beauty of Python is that we do not necessarily need to create four objects (`fp`, `files`,  `bag_o_cubes`, `bag_cubelist`) to get a `CubeList` object but can chain the functions together into one big function call:\n",
+    "        \n",
+    "        iris.cube.CubeList(db.from_sequence(files).map(iris.load_cube).compute())"
```
Comment (dkillick) [inline course_content/notebooks/dask_intro.ipynb]: In this particular case it doesn't actually make any difference (as there's no special syntax to highlight), but in you can also make a preformatted code block as follows that includes syntax highlighting:
```
\```python
iris.cube.CubeList(db.from_sequence(files).map(iris.load_cube).compute())
\```
```

(But drop the leading slashes – I can't quite get the syntax highlighting working correctly, ironically enough.)

File: course_content/notebooks/dask_intro.ipynb [hunk 2]
```diff
@@ -393,6 +393,281 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "merged_cube = iris.load(files).merge_cube()\n",
+    "\n",
+    "print(merge_cube)"
```
Comment (dkillick) [inline course_content/notebooks/dask_intro.ipynb]: You've an inconsistency here – `merged_cube` on L1 but `merge_cube` on L3. My preference is for `merged_cube`.

File: course_content/notebooks/dask_intro.ipynb [hunk 3]
```diff
@@ -393,6 +393,281 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Compare this to the ouput of printing a single cube and it is clear that the decade's worth of data has been merged into a single cube with 120 time steps:"
```
Comment (dkillick) [inline course_content/notebooks/dask_intro.ipynb]: Typo: "ouput"

File: course_content/notebooks/dask_intro.ipynb [hunk 4]
```diff
@@ -393,6 +393,281 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@delayed\n",
+    "def del_merge(cubes):\n",
```
Comment (dkillick) [inline course_content/notebooks/dask_intro.ipynb]: The contents of this code cell are excellent. I wonder whether we should rename `del_*` to `delayed_*` for clarity though, as `del` in Python already has a clearly defined meaning as the delete operation (which is what I immediately thought was going on here).

File: course_content/notebooks/dask_intro.ipynb [hunk 5]
```diff
@@ -393,6 +393,281 @@
...
+    "del_merged_cube = del_merge(bag_o_cubes)\n",
+    "\n",
+    "del_merged_cube.visualize()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Once again if you expand the task graph (double-click image) you can see that `merge_cube()` is now part of it. Compare this to the task graph of `bag-o-cubes`. What is different?\n",
```
Comment (dkillick) [inline course_content/notebooks/dask_intro.ipynb]: Nice – a mini-exercise embedded in the main content. Could we perhaps add a blank cell below for the learner to add their answer?

**Comment by dkillick** [comment]
> in Iris you can simply do `iris.load(files)` and get the same result with shorter and simpler code

Except for the fact that this is not true of course, as in this case Iris will try and merge to the fewest number of cubes possible.

Good! Means we don't need to think about changing that bit and so lose the nice consistency we had.

File: course_content/notebooks/dask_intro.ipynb [hunk 6]
```diff
@@ -393,6 +393,281 @@
...
+    "bag_merged_cube = bag_cubelist.merge_cube()\n",
+    "\n",
+    "print(bag_merged_cube)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "However the `merge_cube` step in this does not leverage Dask. We can utilise Dask in the merge step by decorating a call to `merge_cube` with the `@delayed` decorator. This allows us to integrate it with our task graph:"
```
Comment (pp-mo) [inline course_content/notebooks/dask_intro.ipynb]: I don't quite understand the point of this.
What is delayed actually *doing* for the user in this case?
I can see no opportunity for parallelising the actual merge operation (the calculation), as that only operates on the metadata, and the data part of the merge (aka stacking) is a lazy operation anyway.

**Comment by dkillick** [comment]
In order to move ahead with this I'm going to merge this to bank the excellent update we have here and submit follow-up changes that relate to the comments made above.

File: course_content/notebooks/dask_intro.ipynb [hunk 7]
```diff
@@ -393,6 +393,281 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import iris\n",
+    "import os\n",
+    "import glob\n",
+    "from dask import bag as db"
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb [hunk 8]
```diff
@@ -393,6 +393,281 @@
...
+   "outputs": [],
+   "source": [
+    "bag_o_cubes.visualize()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`bag_o_cubes` is a delayed/\"lazy\" object which we can `visualize` as a task graph and doesn't execute the computation of mapping filenames onto `iris.load_cube()` until we give it the command to do so (namely `bag_o_cubes.compute()`). You can see from the task graph (double-click to enlarge) that each file is being loaded into a cube invidually and in parallel.\n",
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb [hunk 9]
```diff
@@ -393,6 +393,281 @@
...
+    "print('Earliest:    {}\\nLatest:      {}'.format(files[0], files[-1]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Load multiple files into multiple cubes using `iris.load_cube`\n",
+    "\n",
+    "When using Iris to load data that data is stored in a structure called a `cube`. Multiple cubes can be stored in a list called a `CubeList`. We can load the data from our 120 PP files into a `CubeList` by `map`ping our list of filenames (`files`) onto the `iris.load_cube()` function:"
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb [hunk 10]
```diff
@@ -393,6 +393,281 @@
...
+    "type(cubelist)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Load multiple files into multiple cubes using `dask.bag`\n",
+    "\n",
+    "We can parallelise the loading of data from files into cubes by creating a `dask.bag` object which `map`s our list of filenames onto the function `iris.load_cube()`:"
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb (hunk 1 above)
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb (hunk 2 above)
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb (hunk 3 above)
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb (hunk 4 above)
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

File: course_content/notebooks/dask_intro.ipynb (hunk 5 above)
Comment (pp-mo) [reply course_content/notebooks/dask_intro.ipynb]: #111 addresses :white_check_mark:

**Pull request merged**